src/python/pants/backend/java/compile/javac.py

```diff
@@ -238,9 +239,16 @@ async def javac_check(request: JavacCheckRequest) -> CheckResults:
         CoarsenedTargets, Addresses(field_set.address for field_set in request.field_sets)
     )
 
+    resolves = await MultiGet(
```

Comment:
it wasn't 100% clear to me whether I should be calculating resolves for each `CoarsenedTarget` or for all of them (as is done in the compile rule). If this is correct, I'll turn this behaviour into a separate rule, as I've already done this pattern twice in this PR

Reply:
A different resolve for each `CoarsenedTarget` would mean that an end product (e.g, `deploy_jar`) could be built with different classpathes which is not the expected behavior. I believe the point of `CoarsenedTarget` is to handle cycles in the build graph, but ultimately all of the Java files contributing to an end product must be built with the same classpath.

Reply:
@tdyas so, to confirm, I should be computing the full set of targets and building a resolve of that?

Reply:
To clarify, does this PR or the existing code allow different resolves to be built into the same end product?

Reply:
I'm not 100% across what each `coarsened_target` represents, but, the existing code would allow for different resolves, whereas this code would probably allow for fewer different resolves.

I think changing to a single resolve would probably make sense here.

Reply:
Yes, a single resolve should be used for "an entire compilation graph", and it should be selected at the root before that compile is kicked off.

Imagining a situation where you have different resolves for two root targets (`deploy_jar`s for example): even though those binaries might depend on the same inner nodes in the graph, they will each trigger compilation with different resolves. That will mean that the compiles for those distinct resolves will happen independently and in parallel (since the resolve that is being used is part of the request, and thus part of the `@rule`'s identity.)

Reply:
As clarified elsewhere: these lines are in `check`, and `check` is selecting the roots to compile with. So it _should_ select a distinct resolve per coarsened target (the same as `package` and `test` should in the future).

It's only once a compile graph is kicked off that we should be preserving the resolve that was selected.